### PR TITLE
Backport parser tests and some changes from Jedi branch

### PR DESCRIPTION
**Pull request by Nurdok**
Since the Jedi team are working on an API, I'm pausing work on #240. However, I have rewritten the parser tests so I am adding them to pydocstyle in this PR. There are also some changes such as raising `ParseError` is the Python file's syntax is illegal. When this is done, a new version can be released (2.0.0).

#### To Do
- ~~Add a test that runs pydocstyle on two files and make sure they are both checked even if the first has illegal syntax.~~

### CODE REVIEW

File: src/pydocstyle/cli.py [hunk 1]
```diff
@@ -49,7 +49,8 @@ def run_pydocstyle():
 
     count = 0
     for error in errors:
-        sys.stdout.write('%s\n' % error)
+        if hasattr(error, 'code'):
```
Comment (farmersez) [inline src/pydocstyle/cli.py]: I'd check that the type of `error` does not inherit from `BaseException`. Ducktyping here seems... odd (imo).

File: src/pydocstyle/parser.py [hunk 2]
```diff
@@ -30,7 +29,12 @@ def next(obj, default=nothing):
 
 __all__ = ('Parser', 'Definition', 'Module', 'Package', 'Function',
            'NestedFunction', 'Method', 'Class', 'NestedClass', 'AllError',
-           'StringIO')
+           'StringIO', 'ParseError')
+
+
+class ParseError(Exception):
+    def __str__(self):
+        return "cannot parse file"
```
Comment (farmersez) [inline src/pydocstyle/parser.py]: `AllError`'s message is capitalized, so should this.

File: src/tests/parser_test.py [hunk 3]
```diff
@@ -291,3 +429,138 @@ def test_module_publicity():
...
     assert module.is_public
+
+
+def test_complex_module():
+    """Test that a complex module is parsed correctly."""
+    parser = Parser()
+    code = CodeSnippet('''\
+        """Module."""
+        __all__ = ('a', 'b'
+                   'c',)
```
Comment (farmersez) [inline src/tests/parser_test.py]: Is this on a separate line on purpose?

File: src/tests/parser_test.py [hunk 4]
```diff
@@ -291,3 +429,138 @@ def test_module_publicity():
...
+        def function():
+            "Function."
+            def nested_1():
+                """Nested."""
+            if True:
+                def nested_2():
+                    pass
+        class class_(object):
```
Comment (farmersez) [inline src/tests/parser_test.py]: `class_1`? Just an underscore at the end looks fugly 😋

File: src/tests/test_integration.py [hunk 5]
```diff
@@ -954,3 +964,15 @@ def foo():
     _, _, code = env.invoke()
 
     assert code == 0
+
+
+def test_syntax_error_multiple_files(env):
```
Comment (farmersez) [inline src/tests/test_integration.py]: Perhaps add a third _valid_ file and see that it's parsed correctly? (or is that TODO in the issue? Why not do it now?)

File: src/pydocstyle/cli.py (hunk 1 above)
Comment (sigmavirus24) [inline src/pydocstyle/cli.py]: Why? What does checking for that prevent?

File: src/tests/parser_test.py [hunk 6]
```diff
@@ -291,3 +429,138 @@ def test_module_publicity():
...
     module = parser.parse(code, "__filepath")
     assert module.is_public
+
+
+def test_complex_module():
+    """Test that a complex module is parsed correctly."""
+    parser = Parser()
+    code = CodeSnippet('''\
+        """Module."""
+        __all__ = ('a', 'b'
```
Comment (sigmavirus24) [inline src/tests/parser_test.py]: Should there be a trailing comma here?

File: src/tests/test_integration.py (hunk 5 above)
Comment (sigmavirus24) [inline src/tests/test_integration.py]: Well if a single syntax error prevented further checking, then wouldn't it hold that we would never see the second one produce a syntax error either? I don't think a third file is needed here.

File: src/tests/parser_test.py (hunk 3 above)
Comment (Nurdok) [inline src/tests/parser_test.py]: Yes, it's a corner case for the parser.

File: src/tests/test_integration.py (hunk 5 above)
Comment (Nurdok) [inline src/tests/test_integration.py]: What @sigmavirus24 said :)

File: src/tests/parser_test.py (hunk 6 above)
Comment (Nurdok) [inline src/tests/parser_test.py]: No, it's a concatenated string.

File: src/pydocstyle/cli.py (hunk 1 above)
Comment (Nurdok) [inline src/pydocstyle/cli.py]: @farmersez why is `isinstance` better? I think this clearly reads as "print any error that has an error code associated with it".

File: src/pydocstyle/parser.py (hunk 2 above)
Comment (Nurdok) [inline src/pydocstyle/parser.py]: Done.

File: src/tests/parser_test.py (hunk 4 above)
Comment (Nurdok) [inline src/tests/parser_test.py]: It's just test code in a string... I don't think it matters.

File: src/pydocstyle/cli.py (hunk 1 above)
Comment (farmersez) [inline src/pydocstyle/cli.py]: `errors` contain 2 object types - Exceptions (caught by pydocstyle) and errors (generated by pydocstyle). If some day `errors` would contain more than those 2 object types the `hasattr` check might not be good enough, but my proposal might not be good as well. Leave it as it is.
